### PR TITLE
add sidekiq support

### PR DESCRIPTION
**Pull request by troter**
Integration test work well with sidekiq.

### CODE REVIEW

**Comment by isaacseymour** [comment]
Hi @troter - awesome work on this! The reason we didn't put sidekiq support in from the start is because it has retrying built in, so didn't want to risk conflicting with that and over retrying jobs. Do you know how this will interact with sidekiq's automatic retry logic?

**Comment by troter** [comment]
> so didn't want to risk conflicting with that and over retrying jobs.

Yes. if use sidekiq(and these changes), we should remove Sidekiq::Middleware::Server::RetryJobs.
but, this is possible to retry settings for per worker.

> Do you know how this will interact with sidekiq's automatic retry logic?

No, I don't. per worker retry settings define on worker class it included Sidekiq::Worker.
activejob's sidekiq adapter always use `ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper`, and it is difficult to change.

**Comment by isaacseymour** [comment]
Yeah - the job wrapper problem is the same one we ran into. I think actually it shouldn't be a problem - if activejob-retry kicks in the job won't raise an error, so Sidekiq will think it's succeeded and not retry it. If this is right, we can merge this without having to worry about doing stuff to the Sidekiq adapter... @greysteil does this sound right to you?

**Comment by greysteil** [comment]
We'll still raise an error on the final retry if we fail. Without disabling `Sidekiq::Middleware::Server::RetryJobs` that will trigger a retry from the Sidekiq side, which will kick us back into the retry-logic here, no?

**Comment by isaacseymour** [comment]
Ah, yeah, of course. I'm not sure there's a good way of doing this:

- we could `ActiveJob::QueueAdapters::SidekiqAdapter::JobWrapper.sidekiq_options(:retry => false)` when `include ActiveJob::Retry` is run, but that will disable all retries for all jobs (which is quite unexpected)
- we could leave it how it is and expect people using Sidekiq to use `rescue_from(StandardError) { |error| email_to_devs(error) }` in all their activejob-retry controlled jobs

Neither option seems all that nice, although I'm much more in favour of the latter (extra work is better than unexpected behaviour).

:disappointed:

**Comment by greysteil** [comment]
I reckon the later option is fine, as long as it's documented.

File: test/integration/queuing_test.rb [hunk 1]
```diff
@@ -45,8 +45,7 @@ class QueuingTest < ActiveSupport::TestCase
   end
 
   test 'should call rescue_from only when retries have run out' do
-    TestJob.rescue_from(RuntimeError) { write_to_rescue_file }
```
Comment (isaacseymour) [inline test/integration/queuing_test.rb]: Why is this getting removed?

Comment (isaacseymour) [inline test/integration/queuing_test.rb]: Ah, being moved in to the templated job, nice.

**Comment by isaacseymour** [comment]
:+1: will merge at the same time as #29. Thanks for your work on this @troter !

**Pull request merged**